File: app/src/main/java/org/phenoapps/intercross/AuxValueInputActivity.java
package org.phenoapps.intercross;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.zebra.sdk.comm.BluetoothConnection;
import com.zebra.sdk.comm.ConnectionException;
import com.zebra.sdk.graphics.internal.ZebraImageAndroid;
import com.zebra.sdk.printer.PrinterStatus;
import com.zebra.sdk.printer.SGD;
import com.zebra.sdk.printer.ZebraPrinter;
import com.zebra.sdk.printer.ZebraPrinterFactory;
import com.zebra.sdk.printer.ZebraPrinterLanguageUnknownException;
import com.zebra.sdk.printer.ZebraPrinterLinkOs;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AuxValueInputActivity extends AppCompatActivity {

    final static private String line_separator = System.getProperty("line.separator");

    private String mTimestamp;
    private String mCrossId;

    private IdEntryDbHelper mDbHelper;

    private SharedPreferences.OnSharedPreferenceChangeListener mPrefListener;

    private ArrayList<AdapterEntry> mCrossIds;

    private ActionBarDrawerToggle mDrawerToggle;

    private View focusedTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        String crossId = getIntent().getStringExtra("crossId");
       // String[] headers = getIntent().getStringArrayExtra("headers");

        SharedPreferences prefs = getSharedPreferences("Settings", Context.MODE_PRIVATE);

        mCrossId = crossId;

        String[] headers = prefs.getStringSet(SettingsActivity.HEADER_SET, new HashSet<>()).toArray(new String[] {});
        RecyclerView recyclerView = new RecyclerView(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        mDbHelper = new IdEntryDbHelper(this);

        SQLiteDatabase db = mDbHelper.getWritableDatabase();
        HashMap<String, String> colMap = new HashMap<>();

        try {
            String table = IdEntryContract.IdEntry.TABLE_NAME;

            Cursor cursor = db.query(table, headers, "cross_id=?", new String[] {crossId}, null, null, null);

            if (cursor.moveToFirst()) {
                do {
                    final String[] headerCols = cursor.getColumnNames();

                    for (String header : headerCols) {


                        final String val = cursor.getString(
                                cursor.getColumnIndexOrThrow(header)
                        );

                        if (header.equals("timestamp")) mTimestamp = val;


                        colMap.put(header, val);

                    }

                } while (cursor.moveToNext());

            }
            cursor.close();

        } catch (SQLiteException e) {
            e.printStackTrace();
        }

        List<EditTextAdapterEntry> entries = new ArrayList<>();
        List<EditText> editTexts = new ArrayList<>();

        for (String header : headers) {
            EditText editText = new EditText(this);
            editText.setText(colMap.get(header));
            entries.add(new EditTextAdapterEntry(editText, header));
            editTexts.add(editText);
        }

        EditTextRecyclerViewAdapter adapter = new EditTextRecyclerViewAdapter(this, entries);
        recyclerView.setAdapter(adapter);

        LinearLayout view = new LinearLayout(this);
        view.setOrientation(LinearLayout.VERTICAL);
        view.addView(recyclerView);

        Button submitButton = new Button(this);// = entry.editText;
        submitButton.setText("Update");
        submitButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int next = 0;

                for (String header: headers) {
                    db.execSQL("UPDATE INTERCROSS SET " +
                            header + " = '" + editTexts.get(next++).getText().toString() + //adapter.getItem(next++).editText.getText().toString() + //entries.get(next++).editText.getText() +
                            "' WHERE cross_id = '" + crossId + "'");
                }
            }
        });

        view.addView(submitButton);
        setContentView(view);

    }

    @Override
    final public boolean onCreateOptionsMenu(Menu m) {

        final MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.drawer_print_layout, m);
        return true;
    }

    @Override
    final public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.action_print:
                BluetoothAdapter mBluetoothAdapter = null;
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.ECLAIR) {
                    mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
                    Set<BluetoothDevice> pairedDevices = mBluetoothAdapter.getBondedDevices();

                    //TODO allow multiple pairs
                    //TODO wrap in async task
                    if (pairedDevices.size() == 1) {
                        BluetoothDevice bd = pairedDevices.toArray(new BluetoothDevice[] {})[0];
                        Log.d("BT", "PAIRED");
                        BluetoothConnection bc = new BluetoothConnection(bd.getAddress());
                        try {
                            bc.open();
                            final ZebraPrinter printer = ZebraPrinterFactory.getInstance(bc);
                            ZebraPrinterLinkOs linkOsPrinter = ZebraPrinterFactory.createLinkOsPrinter(printer);
                            PrinterStatus printerStatus = (linkOsPrinter != null) ? linkOsPrinter.getCurrentStatus() : printer.getCurrentStatus();
                            getPrinterStatus(bc);
                            if (printerStatus.isReadyToPrint) {
                                runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {

                                        Toast.makeText(AuxValueInputActivity.this, "Printer Ready", Toast.LENGTH_LONG).show();

                                    }
                                });

                                //printer.sendCommand("! DF RUN.BAT ! UTILITIES JOURNAL SETFF 50 5 PRINT");
                                //printer.printConfigurationLabel();
                                //printer.sendCommand("^XA^FO0,0^ADN,36,20^FDCHANEY^FS^XZ");
                                printer.sendCommand("^XA"
                                        + "^FWR"
                                        + "^FO100,75^A0,25,20^FD" + mCrossId + "^FS"
                                        + "^FO200,75^A0N,25,20"
                                        + "^BQN,2,10^FDMA" + mCrossId + "^FS"
                                        + "^FO450,75^A0,25,20^FD" + mTimestamp + "^FS^XZ");
                                /*printer.printImage(new ZebraImageAndroid(BitmapFactory.decodeResource(getApplicationContext().getResources(),
                                        R.drawable.intercross_small)), 75,500,-1,-1,false);*/

                            } else if (printerStatus.isHeadOpen) {
                                //helper.showErrorMessage("Error: Head Open \nPlease Close Printer Head to Print");
                            } else if (printerStatus.isPaused) {
                                //helper.showErrorMessage("Error: Printer Paused");
                            } else if (printerStatus.isPaperOut) {
                                //helper.showErrorMessage("Error: Media Out \nPlease Load Media to Print");
                            } else {
                                //helper.showErrorMessage("Error: Please check the Connection of the Printer");
                            }

                            bc.close();

                        } catch (ConnectionException e) {
                            e.printStackTrace();
                        } catch (ZebraPrinterLanguageUnknownException e) {
                            e.printStackTrace();
                        }
                    }
                }
                break;
            default:
                return super.onOptionsItemSelected(item);
        }
        return true;
    }

    private void getPrinterStatus(BluetoothConnection connection) throws ConnectionException{


        final String printerLanguage = SGD.GET("device.languages", connection); //This command is used to get the language of the printer.

        final String displayPrinterLanguage = "Printer Language is " + printerLanguage;

        SGD.SET("device.languages", "zpl", connection); //This command set the language of the printer to ZPL

        AuxValueInputActivity.this.runOnUiThread(new Runnable() {
            @Override
            public void run() {

                Toast.makeText(AuxValueInputActivity.this,
                        displayPrinterLanguage + "\n" + "Language set to ZPL", Toast.LENGTH_LONG).show();

            }
        });

    }

}

File: app/src/main/java/org/phenoapps/intercross/AuxValueInputActivity.kt
package org.phenoapps.intercross

import android.bluetooth.BluetoothAdapter
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.preference.PreferenceManager
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.zebra.sdk.comm.BluetoothConnection
import com.zebra.sdk.comm.ConnectionException
import com.zebra.sdk.printer.SGD
import com.zebra.sdk.printer.ZebraPrinterFactory
import com.zebra.sdk.printer.ZebraPrinterLanguageUnknownException
import org.w3c.dom.Text

//TODO create separate file for async bluetooth task
class AuxValueInputActivity : AppCompatActivity() {

    private val mIdTextView: TextView by lazy {
        findViewById<TextView>(R.id.selfTextView)
    }
    private val mNotBpParentTextView: TextView by lazy {
        findViewById<TextView>(R.id.nbpTextView)
    }
    private val mDateTextView: TextView by lazy {
        findViewById<TextView>(R.id.dateTextView)
    }
    private val mPolTypeTextView: TextView by lazy {
        findViewById<TextView>(R.id.polTypeTextView)
    }
    private val mBpMaleTextView: TextView by lazy {
        findViewById<TextView>(R.id.bpMaleTextView)
    }
    private val mBpFemaleTextView: TextView by lazy {
        findViewById<TextView>(R.id.bpFemaleTextView)
    }
    private val mPersonTextView: TextView by lazy {
        findViewById<TextView>(R.id.personTextView)
    }
    private val mRecyclerView: RecyclerView by lazy {
        findViewById<RecyclerView>(R.id.offspringRecyclerView)
    }

    private val mEntries = ArrayList<AdapterEntry>()

    private val mDbHelper = IdEntryDbHelper(this)

    private var mMaleParent = String()
    private var mFemaleParent = String()
    private var mCrossId = String()
    private var mTimestamp = String()
    private var mPerson = String()

    private var mZplFileName = String()
    private val mCode: String by lazy {
        PreferenceManager.getDefaultSharedPreferences(this)
                .getString("ZPL_CODE", "")
    }

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_manage_values)

        supportActionBar?.let {
            it.title = ""
            it.themedContext
            it.setDisplayHomeAsUpEnabled(true)
            it.setHomeButtonEnabled(true)
        }

        val id = intent.getIntExtra(IntercrossConstants.COL_ID_KEY, -1)
        mCrossId = intent.getStringExtra(IntercrossConstants.CROSS_ID) ?: ""
        mMaleParent = intent.getStringExtra(IntercrossConstants.MALE_PARENT) ?: ""
        mFemaleParent = intent.getStringExtra(IntercrossConstants.FEMALE_PARENT) ?: ""
        mTimestamp = mDbHelper.getTimestampById(id)
        mPerson = mDbHelper.getPersonById(id)

        mIdTextView.text = "Cross ID: $mCrossId"
        mDateTextView.text = "$mTimestamp"
        mPersonTextView.text = "by $mPerson"

        if (mCode.isNotBlank()) {
            if (mCode.contains("DFR:")) {
                val split = mCode.split("DFR:")
                if (split.size > 1) {
                    val second = split[1].split("^FS")
                    if (second.size > 1) {
                        mZplFileName = second[0]
                    }
                }
            }
        }

        val polType = mDbHelper.getPollinationType(id)
        mPolTypeTextView.text = polType

        when (polType) {
            "Biparental" -> {
                mBpFemaleTextView.visibility = View.VISIBLE
                mBpMaleTextView.visibility = View.VISIBLE
                mNotBpParentTextView.visibility = View.INVISIBLE
                mBpMaleTextView.text = mMaleParent
                mBpFemaleTextView.text = mFemaleParent
                mBpMaleTextView.setOnClickListener {
                    val maleId = mDbHelper.getRowId(mMaleParent)
                    if (maleId != -1) startCrossActivity(id, mMaleParent)
                    else Toast.makeText(this, "This id has not DB entry.", Toast.LENGTH_SHORT).show()
                }
                mBpFemaleTextView.setOnClickListener {
                    val femaleId = mDbHelper.getRowId(mFemaleParent)
                    if (femaleId != -1) startCrossActivity(mDbHelper.getRowId(mFemaleParent), mFemaleParent)
                    else Toast.makeText(this, "This id has not DB entry.", Toast.LENGTH_SHORT).show()
                }
            }
            "OpenPollinated", "SelfPollinated" -> {
                mBpFemaleTextView.visibility = View.INVISIBLE
                mBpMaleTextView.visibility = View.INVISIBLE
                mNotBpParentTextView.visibility = View.VISIBLE
                mNotBpParentTextView.text = mFemaleParent
                mNotBpParentTextView.setOnClickListener {
                    val femaleId = mDbHelper.getRowId(mFemaleParent)
                    if (femaleId != -1) startCrossActivity(mDbHelper.getRowId(mFemaleParent), mFemaleParent)
                    else Toast.makeText(this, "This id has not DB entry.", Toast.LENGTH_SHORT).show()
                }
            }
        }

        val offspring = mDbHelper.getOffspring(id)

        mRecyclerView.layoutManager = LinearLayoutManager(this)

        offspring.forEach { child ->
            mEntries.add(AdapterEntry(child.first, child.second))
        }

        val adapter = object : ViewAdapter<AdapterEntry>(mEntries) {

            override fun getLayoutId(position: Int, obj: AdapterEntry): Int {
                return R.layout.row
            }

            override fun getViewHolder(view: View, viewType: Int): RecyclerView.ViewHolder {
                return ViewHolder(view)
            }
        }

        mRecyclerView.adapter = adapter
    }

    inner class ViewHolder internal constructor(itemView: View) :
            RecyclerView.ViewHolder(itemView), ViewAdapter.Binder<AdapterEntry> {

        private var firstText: TextView = itemView.findViewById(R.id.crossTextView) as TextView

        private var mEntry: AdapterEntry = AdapterEntry()

        init {
            firstText.setOnClickListener {
                startCrossActivity(mEntry.second.toInt(), mEntry.first)
            }
        }

        override fun bind(data: AdapterEntry) {
            mEntry = data
            firstText.text = data.first
        }
    }

    private fun startCrossActivity(id: Int, crossId: String) {

        val pref = PreferenceManager.getDefaultSharedPreferences(this@AuxValueInputActivity)

        val parents = mDbHelper.getParents(id)

        val timestamp = mDbHelper.getTimestampById(id)

        val intent = Intent(this@AuxValueInputActivity, AuxValueInputActivity::class.java)

        intent.putExtra(IntercrossConstants.COL_ID_KEY, id)

        intent.putExtra(IntercrossConstants.CROSS_ID, crossId)

        intent.putExtra(IntercrossConstants.TIMESTAMP, timestamp)

        intent.putExtra(IntercrossConstants.FEMALE_PARENT, parents[0])

        intent.putExtra(IntercrossConstants.MALE_PARENT, parents[1])

        intent.putExtra(IntercrossConstants.PERSON, pref.getString(SettingsActivity.PERSON, ""))

        startActivityForResult(intent, IntercrossConstants.USER_INPUT_HEADERS_REQ)
    }

    override fun onCreateOptionsMenu(m: Menu): Boolean {

        val inflater = menuInflater
        inflater.inflate(R.menu.drawer_print_layout, m)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        when (item.itemId) {
            R.id.action_print -> {

                object : AsyncTask<Void, Void, String>() {

                    override fun doInBackground(vararg params: Void?): String {

                        val pref = PreferenceManager.getDefaultSharedPreferences(this@AuxValueInputActivity)
                        val btId = pref.getString(SettingsActivity.BT_ID, "")

                        if (btId.isBlank()) {
                            run { Toast.makeText(this@AuxValueInputActivity,
                                    "No bluetooth device paired.", Toast.LENGTH_SHORT) }
                        } else {
                            val mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter()

                            val pairedDevices = mBluetoothAdapter.bondedDevices.filter {
                                it.name == btId
                            }

                            if (pairedDevices.isNotEmpty()) {
                                val bc = BluetoothConnection(pairedDevices[0].address)

                                try {
                                    bc.open()
                                    val printer = ZebraPrinterFactory.getInstance(bc)
                                    val linkOsPrinter = ZebraPrinterFactory.createLinkOsPrinter(printer)
                                    linkOsPrinter?.let {
                                        val printerStatus = it.currentStatus
                                        getPrinterStatus(bc)
                                        if (printerStatus.isReadyToPrint) {

                                            if (mCode.isNotBlank() && mZplFileName.isNotBlank()) {
                                                printer.sendCommand(mCode)
                                                printer.sendCommand("^XA"
                                                        + "^XFR:${mZplFileName}"
                                                        + "^FN1^FD" + mCrossId + "^FS"
                                                        + "^FN2^FDQA," + mCrossId + "^FS"
                                                        + "^FN3^FD" + mTimestamp + "^FS^XZ")
                                            } else {
                                                printer.sendCommand("^XA"
                                                        + "^MNA"
                                                        + "^MMT,N"
                                                        + "^DFR:DEFAULT_INTERCROSS_SAMPLE.GRF^FS"
                                                        + "^FWR"
                                                        + "^FO100,25^A0,25,20^FN1^FS"
                                                        + "^FO200,25^A0N,25,20"
                                                        + "^BQ,2,6" +
                                                        "^FN2^FS"
                                                        + "^FO450,25^A0,25,20^FN3^FS^XZ")

                                                printer.sendCommand("^XA"
                                                        + "^XFR:DEFAULT_INTERCROSS_SAMPLE.GRF"
                                                        + "^FN1^FD" + mCrossId + "^FS"
                                                        + "^FN2^FDQA," + mCrossId + "^FS"
                                                        + "^FN3^FD" + mTimestamp + "^FS^XZ")
                                            }
                                            /*printer.printImage(new ZebraImageAndroid(BitmapFactory.decodeResource(getApplicationContext().getResources(),
                                    R.drawable.intercross_small)), 75,500,-1,-1,false);*/

                                        } else if (printerStatus.isHeadOpen) {
                                            runOnUiThread { Toast.makeText(this@AuxValueInputActivity, "Printer is open.", Toast.LENGTH_LONG).show() }
                                        } else if (printerStatus.isPaused) {
                                            runOnUiThread { Toast.makeText(this@AuxValueInputActivity, "Printer is paused.", Toast.LENGTH_LONG).show() }
                                        } else if (printerStatus.isPaperOut) {
                                            runOnUiThread { Toast.makeText(this@AuxValueInputActivity, "No paper.", Toast.LENGTH_LONG).show() }
                                        } else {
                                            runOnUiThread { Toast.makeText(this@AuxValueInputActivity, "Please check the printer's connection.", Toast.LENGTH_LONG).show() }
                                        }
                                    }
                                } catch (e: ConnectionException) {
                                    e.printStackTrace()
                                } catch (e: ZebraPrinterLanguageUnknownException) {
                                    e.printStackTrace()
                                } finally {
                                    bc.close()
                                }
                            }
                        }

                        return String()
                    }

                }.execute()


            }
            else -> finish()
        }
        return true
    }

    @Throws(ConnectionException::class)
    private fun getPrinterStatus(connection: BluetoothConnection) {

        val printerLanguage = SGD.GET("device.languages", connection) //This command is used to get the language of the printer.

        val displayPrinterLanguage = "Printer Language is $printerLanguage"

        SGD.SET("device.languages", "zpl", connection) //This command set the language of the printer to ZPL

        this@AuxValueInputActivity.runOnUiThread {
            Toast.makeText(this@AuxValueInputActivity,
                    "$displayPrinterLanguage\nLanguage set to ZPL", Toast.LENGTH_LONG).show()
        }
    }

    companion object {

        private val line_separator = System.getProperty("line.separator")
    }

}
